### PR TITLE
[PDI-13875] Updated Google Analytics plugin to v3 APIs, moved to core plugin

### PR DESCRIPTION
**Pull request by mattyb149**
Also removed references to GData client libraries from other PDI classes like Kitchen and LdapProtocol

### CODE REVIEW

File: engine/src/org/pentaho/di/kitchen/Kitchen.java [hunk 1]
```diff
@@ -77,16 +78,16 @@ public static void main( String[] a ) throws Exception {
     final ExecutorService executor = ExecutorUtil.getExecutor();
     final RepositoryPluginType repositoryPluginType = RepositoryPluginType.getInstance();
 
-    final Future<Pair<KettlePluginException, Future<KettleException>>> repositoryRegisterFuture =
-      executor.submit( new Callable<Pair<KettlePluginException, Future<KettleException>>>() {
+    final Future<Map<KettlePluginException, Future<KettleException>>> repositoryRegisterFuture =
```
Comment (mbatchelor) [inline engine/src/org/pentaho/di/kitchen/Kitchen.java]: Pair is a simple struct - Map<KettlePluginException, Future<KettleException>> seems like the wrong thing to use. I suggest:

Map.Entry<KettlePluginException, Future<KettleException>> - - - Map.Entry is a very similar structure to Pair.

File: plugins/googleanalytics/src/org/pentaho/di/trans/steps/googleanalytics/GaInputStep.java [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+      }
+    }
+
+    // generate output row, make it correct size
+    Object[] outputRow = RowDataUtil.allocateRowData( data.outputRowMeta.size() );
+
+    List<String> entry = null;
+
+    // no query prepared yet?
+    if ( data.query == null ) {
```
Comment (mbatchelor) [inline plugins/googleanalytics/src/org/pentaho/di/trans/steps/googleanalytics/GaInputStep.java]: Any reason this code was moved from the getNextEntry method? Seems like having it all crammed into the processRow makes processRow less readable.

File: plugins/googleanalytics/src/org/pentaho/di/trans/steps/googleanalytics/GaInputStep.java [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+
+      data.query = getQuery( analytics );
+      // use default max results for now
+      //data.query.setMaxResults( 10000 );
+
```
Comment (mbatchelor) [inline plugins/googleanalytics/src/org/pentaho/di/trans/steps/googleanalytics/GaInputStep.java]: old code had detailed log output here:
      if ( log.isDetailed() ) {
        logDetailed( "querying google analytics: " + data.query.getUrl() );
      }

File: plugins/googleanalytics/src/org/pentaho/di/trans/steps/googleanalytics/GaInputStep.java [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+        }
+      } catch ( IOException e2 ) {
+        throw new KettleException( e2 );
+      }
+    }
+
+    List<List<String>> entries = data.feed.getRows();
+    if ( data.entryIndex < entries.size() ) {
+      entry = entries.get( data.entryIndex++ );
+    }
```
Comment (mbatchelor) [inline plugins/googleanalytics/src/org/pentaho/di/trans/steps/googleanalytics/GaInputStep.java]: Shouldn't there be an : 
else {
  entry = null;
}

Old getNextEntry method returned null indicating no more entries...

File: plugins/googleanalytics/src/org/pentaho/di/trans/steps/googleanalytics/GaInputStep.java [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+        environmentSubstitute( meta.getMetrics() ) // metrics
+      );
+      query.setDimensions( environmentSubstitute( meta.getDimensions() ) );
+
+
+      if ( meta.isUseSegment() ) {
+        if ( meta.isUseCustomSegment() ) {
+          query.setSegment( environmentSubstitute( meta.getCustomSegment() ) );
+        } else {
+          query.setSegment( meta.getSegmentId() );
```
Comment (mbatchelor) [inline plugins/googleanalytics/src/org/pentaho/di/trans/steps/googleanalytics/GaInputStep.java]: environmentSubstitute here ?

File: plugins/googleanalytics/src/org/pentaho/di/trans/steps/googleanalytics/GaInputStep.java [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+            GaInputStepMeta.DEPRECATED_FIELD_TYPE_CONFIDENCE_INTERVAL ) );
+        }
+      }
+    }
+
+    try {
+      // Create an Analytics object, and fetch what we can for later (account name, e.g.)
+      analytics = GoogleAnalyticsApiFacade.createFor( environmentSubstitute( meta.getGaAppName() ), meta )
+        .getAnalytics();
+      accountName = analytics.management().accounts().list().execute().getItems().iterator().next().getName();
```
Comment (mbatchelor) [inline plugins/googleanalytics/src/org/pentaho/di/trans/steps/googleanalytics/GaInputStep.java]: This is a lot of .x.x.x - - any chance of an NPE blowing this up ?

* Can accounts() return null?
* Can execute() return null?
* Can getItems() return null?

File: plugins/googleanalytics/src/org/pentaho/di/trans/steps/googleanalytics/GaInputStep.java (hunk 5 above)
Comment (mattyb149) [inline plugins/googleanalytics/src/org/pentaho/di/trans/steps/googleanalytics/GaInputStep.java]: The else clause is when you've clicked "Get Segments" and have chosen from a dropdown list

File: plugins/googleanalytics/src/org/pentaho/di/trans/steps/googleanalytics/GaInputStep.java [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+
+    List<List<String>> entries = data.feed.getRows();
+    if ( data.entryIndex < entries.size() ) {
+      entry = entries.get( data.entryIndex++ );
+    }
+
+    if ( entry != null && ( meta.getRowLimit() <= 0 || getLinesWritten() < meta.getRowLimit() ) ) { // another record to
+      // fill the output fields with look up data
+      for ( int i = 0, j = 0; i < meta.getFeedField().length; i++ ) {
+        String fieldName = meta.getFeedField()[i];
```
Comment (mbatchelor) [inline plugins/googleanalytics/src/org/pentaho/di/trans/steps/googleanalytics/GaInputStep.java]: Old step used environmentSubstitute - I believe you need to as well: 
String fieldName = environmentSubstitute( meta.getFeedField()[i] );

File: plugins/googleanalytics/src/org/pentaho/di/trans/steps/googleanalytics/GaInputStep.java (hunk 2 above)
Comment (mattyb149) [inline plugins/googleanalytics/src/org/pentaho/di/trans/steps/googleanalytics/GaInputStep.java]: Originally I thought I needed access to the query and the row, but it turned out I didn't. Will move it back into the method

File: plugins/googleanalytics/src/org/pentaho/di/trans/steps/googleanalytics/GaInputStep.java (hunk 6 above)
Comment (mattyb149) [inline plugins/googleanalytics/src/org/pentaho/di/trans/steps/googleanalytics/GaInputStep.java]: If any return null, we're screwed, so the catch block will get the NPE and return false

File: plugins/googleanalytics/src/org/pentaho/di/trans/steps/googleanalytics/GoogleAnalyticsApiFacade.java [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    Assert.assertNotNull( keyFile, "OAuth secret key file cannot be null" );
+
+    this.httpTransport = httpTransport;
+
+    Credential credential = new GoogleCredential.Builder()
+      .setTransport( httpTransport )
+      .setJsonFactory( jsonFactory )
+      .setServiceAccountScopes( AnalyticsScopes.all() )
+      .setServiceAccountId( oathServiceEmail )
+      .setServiceAccountPrivateKeyFromP12File( keyFile )
```
Comment (mbatchelor) [inline plugins/googleanalytics/src/org/pentaho/di/trans/steps/googleanalytics/GoogleAnalyticsApiFacade.java]: We need to put in an improvement to support:
setServiceAccountPrivateKey(PrivateKey serviceAccountPrivateKey)
and
setServiceAccountPrivateKeyFromPemFile(File pemFile)

File: plugins/googleanalytics/src/org/pentaho/di/trans/steps/googleanalytics/GoogleAnalyticsApiFacade.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+  public static GoogleAnalyticsApiFacade createFor( String application, GaInputStepMeta meta )
+    throws IOException, GeneralSecurityException {
+
+    return new GoogleAnalyticsApiFacade(
+      GoogleNetHttpTransport.newTrustedTransport(),
+      JacksonFactory.getDefaultInstance(),
+      application,
+      meta.getOAuthServiceAccount(),
+      new File( meta.getOAuthKeyFile() ) );
```
Comment (mbatchelor) [inline plugins/googleanalytics/src/org/pentaho/di/trans/steps/googleanalytics/GoogleAnalyticsApiFacade.java]: Neet environmentSubstitute here (or somewhere close by) ... as the transformation moves from place to place, the key file will be in different places for sure.

File: plugins/googleanalytics/src/org/pentaho/di/trans/steps/googleanalytics/GoogleAnalyticsApiFacade.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+  private Analytics analytics;
+
+  public static GoogleAnalyticsApiFacade createFor( String application, GaInputStepMeta meta )
+    throws IOException, GeneralSecurityException {
+
+    return new GoogleAnalyticsApiFacade(
+      GoogleNetHttpTransport.newTrustedTransport(),
+      JacksonFactory.getDefaultInstance(),
+      application,
+      meta.getOAuthServiceAccount(),
```
Comment (mbatchelor) [inline plugins/googleanalytics/src/org/pentaho/di/trans/steps/googleanalytics/GoogleAnalyticsApiFacade.java]: Need environmentSubstitute here as well.

**Pull request merged**